community/graphdb-api/src/main/java/org/neo4j/graphdb/PropertyContainer.java

```diff
@@ -145,4 +145,15 @@
      * @return all properties on this property container
      */
     Map<String, Object> getAllProperties();
+
+    /**
+     * Returns the unique id of this <code>PropertyContainer</code>. Ids are garbage collected over time
+     * so they are only guaranteed to be unique during a specific time span: if
+     * the  <code>PropertyContainer</code> is deleted, it's likely that a new
+     * <code>PropertyContainer</code> at some point will get the old id. <b>Note</b>: this makes
+     * <code>PropertyContainer</code> ids brittle as public APIs.
+     *
+     * @return The id of this  <code>PropertyContainer</code>.
+     */
+    long getId();
```

Comment:
Instead of adding a `default` implementation throwing `UnsupportedOperationException`, maybe we can add an extra interface that extends `PropertyContainer` and has the `getId` method. Only `Node`/`Relationship` (and similar) will implement this new interface and we leave untouched the other implementation of `PropertyContainer`.